8347489: RISC-V: Misaligned memory access with COH

RealFYang opened the pull request:
Hi, please consider this change.

We have different base_offset for T_BYTE/T_CHAR (4-byte instead of 8-byte aligned) with COH. This causes misaligned memory accesses for several instrinsics like String.Compare or String.Equals. The reason is that we assume 8-byte alignment and process one 8-byte word starting at the first array element for each iteration in the main loop. As a result, we have performance regressions on platforms with slow misaligned memory accesses like Unmatched and Premier P550 SBCs.

PS: Same issue is there even without COH. base_offset for T_BYTE/T_CHAR is 20 (thus 4-byte aligned) when `UseCompressedClassPointers` is disabled in this case.

Correctness test on linux-riscv64:
- [x] tier1 (TEST_VM_OPTS="-XX:+UnlockExperimentalVMOptions -XX:+UseCompactObjectHeaders") (release)
- [x] tier1 (TEST_VM_OPTS="-XX:+UnlockExperimentalVMOptions -XX:-UseCompactObjectHeaders") (release)
- [x] hotspot:tier1 (TEST_VM_OPTS="-XX:+UnlockExperimentalVMOptions -XX:+UseCompactObjectHeaders") (fastdebug)
- [x] hotspot:tier1 (TEST_VM_OPTS="-XX:+UnlockExperimentalVMOptions -XX:-UseCompactObjectHeaders") (fastdebug)

Performance test on Premier P550 (-XX:+AlwaysPreTouch -Xms8g -Xmx8g):

SPECjbb2005:
```
1. Without Patch
1.1 -XX:+UseParallelGC -XX:-UseCompactObjectHeaders: 32666
1.2 -XX:+UseParallelGC -XX:+UseCompactObjectHeaders: 27610
1.3 -XX:+UseG1GC       -XX:-UseCompactObjectHeaders: 30911
1.4 -XX:+UseG1GC       -XX:+UseCompactObjectHeaders: 26008

2. With Patch
2.1 -XX:+UseParallelGC -XX:-UseCompactObjectHeaders: 32820
2.2 -XX:+UseParallelGC -XX:+UseCompactObjectHeaders: 34179
2.3 -XX:+UseG1GC       -XX:-UseCompactObjectHeaders: 30620
2.4 -XX:+UseG1GC       -XX:+UseCompactObjectHeaders: 31936
```

SPECjbb2015:
```
1. Without Patch
1.1 -XX:+UseG1GC -XX:-UseCompactObjectHeaders: max-jOPS = 1444, critical-jOPS = 431
1.2 -XX:+UseG1GC -XX:+UseCompactObjectHeaders: max-jOPS = 1092, critical-jOPS = 335

2. With Patch
2.1 -XX:+UseG1GC -XX:-UseCompactObjectHeaders: max-jOPS = 1452, critical-jOPS = 419
2.2 -XX:+UseG1GC -XX:+UseCompactObjectHeaders: max-jOPS = 1438, critical-jOPS = 477
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8347489](https://bugs.openjdk.org/browse/JDK-8347489): RISC-V: Misaligned memory access with COH (**Bug** - P4)


### Reviewers
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)
 * [Vladimir Kempik](https://openjdk.org/census#vkempik) (@VladimirKempik - Committer) Review applies to [c093e451](https://git.openjdk.org/jdk/pull/23053/files/c093e451bd4fa4e13ea83308e69cc4cc30be8bae)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23053/head:pull/23053` \
`$ git checkout pull/23053`

Update a local copy of the PR: \
`$ git checkout pull/23053` \
`$ git pull https://git.openjdk.org/jdk.git pull/23053/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23053`

View PR using the GUI difftool: \
`$ git pr show -t 23053`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23053.diff">https://git.openjdk.org/jdk/pull/23053.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23053#issuecomment-2588891343)
</details>

Hamlin-Li reviewed (commented):
Thanks for the patch.
Can you post the performance data in description?
And some minor comments.

RealFYang commented:
> base_offset

Thanks for the suggestions. I have updated accordingly. Note that base_offset is 12 and 16 bytes respectively w/wo COH.
And I have also added some specjbb2005 scores in PR description for reference.

VladimirKempik commented on src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:
```diff
@@ -4443,8 +4463,7 @@ class StubGenerator: public StubCodeGenerator {
       if (multi_block) {
-        int total_adds = vset_sew == Assembler::e32 ? 240 : 608;
-        __ addi(consts, consts, -total_adds);
+        __ subi(consts, consts, vset_sew == Assembler::e32 ? 240 : 608);
```
makes the code less readable, maybe keep total_adds but keep the change of addi to subi ?

RealFYang replied:
Sure. Fixed.

Hamlin-Li commented:
> @Hamlin-Li Thanks for the suggestions. I have updated accordingly. Note that base_offset is 12 and 16 bytes respectively w/wo COH. And I have also added some specjbb2005 scores in PR description for reference.

Thank you posting the data!
As G1 is the default one, can you run the specjbb with G1? And it would be good to list both critial and max value for specjbb.
BTW, there are also some string comparison tests in jmh, e.g. StringCompareToDifferentLength and so on.

Hamlin-Li commented:
> -XX:+UseParallelGC -XX:+AlwaysPreTouch -Xms8g -Xmx8g

I guess the RVV is disabled by default? Will we also address the similar issue when RVV enabled?

Hamlin-Li commented on src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:
```diff
@@ -2461,11 +2461,19 @@ class StubGenerator: public StubCodeGenerator {
+                           (UseCompressedClassPointers ? 8 : 4))) == 0, "Must be");
+
+    // strL is 8-byte aligned
     __ ld(tmpLval, Address(strL));
```
This should also be a `load_long_misaligned`?

Hamlin-Li replied:
And the below one in this function?

RealFYang replied:
Good catch! I think I missed the second one. Fixed.
For the first one, I think we are OK as `strL` is ensured to be 8-byte aligned by the caller [1].

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/stubGenerator_riscv.cpp#L2526

Hamlin-Li commented on src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:
```diff
@@ -2510,17 +2528,19 @@ class StubGenerator: public StubCodeGenerator {
-    __ xorr(tmp3, tmpU, tmpL);
-    __ bnez(tmp3, CALCULATE_DIFFERENCE);
-    __ addi(cnt2, cnt2, -wordSize / 2);
+    if ((base_offset1 % 8) == 0) {
```
The condition should be `((base_offset1 % 8) != 0)`? Seems we try to make the `strL` aligned in the main loop below.

Hamlin-Li replied:
And do we need below piece of code when `AvoidUnalignedAccesses == false`?

RealFYang replied:
The alignment is 8 bytes without COH. And the caller of the stub has loaded 4 characters [1]. So when we need to process another 4 characters to make `strL` 8-byte aligned when entering the main loop. But this is not needed with COH, because the alignment is 4 bytes. That's why I added this if check. Hope that explains.

And I have added check for `AvoidUnalignedAccesses` to this logic.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/stubGenerator_riscv.cpp#L2511

Hamlin-Li reviewed (commented):
Some comments.

RealFYang commented:
> > @Hamlin-Li Thanks for the suggestions. I have updated accordingly. Note that base_offset is 12 and 16 bytes respectively w/wo COH. And I have also added some specjbb2005 scores in PR description for reference.
> 
> Thank you posting the data! As G1 is the default one, can you run the specjbb with G1? And it would be good to list both critial and max value for specjbb. BTW, there are also some string comparison tests in jmh, e.g. StringCompareToDifferentLength and so on.

The purpose of using ParallelGC for the test is to minimize the GC noise. I have also updated SPECjbb2005 score under G1GC in PR description. Note that SPECjbb2005 score means bops (business operations per second). I will update SPECjbb2015 later which has max and critical jops. Thanks.

RealFYang commented:
> > -XX:+UseParallelGC -XX:+AlwaysPreTouch -Xms8g -Xmx8g
> 
> I guess the RVV is disabled by default? Will we also address the similar issue when RVV enabled?

Yes. SBCs like Unmatched and P550 don't have RVV extensions. And I don't see such an issue with RVV code path for these intrinsics.

RealFYang commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1410,6 +1410,14 @@ void C2_MacroAssembler::string_compare(Register str1, Register str2,
+
+  assert((base_offset1 % (UseCompactObjectHeaders ? 4 :
+                          (UseCompressedClassPointers ? 8 : 4))) == 0, "Must be");
+  assert((base_offset2 % (UseCompactObjectHeaders ? 4 :
```
Seems to me that is more than needed. Our code doesn't require that `base_offset1` must equal `base_offset2`. We only care about the alignment of them here, which is reflected on the current two assertions.

RealFYang commented on src/hotspot/cpu/riscv/stubGenerator_riscv.cpp:
```diff
@@ -2488,12 +2496,22 @@ class StubGenerator: public StubCodeGenerator {
+
+    assert((base_offset1 % (UseCompactObjectHeaders ? 4 :
+                            (UseCompressedClassPointers ? 8 : 4))) == 0, "Must be");
+    assert((base_offset2 % (UseCompactObjectHeaders ? 4 :
```
Please see my comment above.

Hamlin-Li commented on src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:
```diff
@@ -1434,8 +1460,14 @@ void C2_MacroAssembler::string_compare(Register str1, Register str2,
     if (str1_isL == str2_isL) { // LL or UU
-      // check if str1 and str2 is same pointer
-      beq(str1, str2, DONE);
+#ifdef ASSERT
```
Can we add the some comment about 8-bytes alignment below at line 1520?
I know it's redundant, but as the code is getting more complicated, it might be good to add it to improve the readability.

RealFYang replied:
Sure. Done.

Hamlin-Li reviewed (approved):
Looks good, thanks for updating.
Just one minor comment.

Hamlin-Li reviewed (approved):
Thank you!

RealFYang commented:
Thanks all for the review!
/integrate
